### PR TITLE
Fix the OE service where returns database as the root node for database connection

### PR DESCRIPTION
**Pull request by AbbiePetcht**

### CODE REVIEW

File: test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectExplorer/ObjectExplorerServiceTests.cs [hunk 1]
```diff
@@ -26,12 +26,26 @@ public class ObjectExplorerServiceTests
...
+            using (SqlTestDb testDb = SqlTestDb.CreateNew(TestServerType.OnPrem, false, databaseName, query, uri))
+            {
+                var session = await CreateSession(null, uri);
+                await CreateSessionAndServerNode(testDb.DatabaseName, session);
+                CancelConnection(uri);
+            }
+        }
+
+        [Fact]
+        public async void CreateSessionAndExpandOnTheDatabaseShouldReturnDatabaseObjects()
```
Comment (llali) [inline test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectExplorer/ObjectExplorerServiceTests.cs]: none of these tests pass master or empty string as database name. can you please add a test to verify that. the code coverage is dropped in this pr so it might be a bug or missing tests

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/ObjectExplorer/ObjectExplorerServiceTests.cs [hunk 2]
```diff
@@ -78,7 +78,13 @@ public ObjectExplorerServiceTests()
         public async Task CreateSessionRequestReturnsSuccessAndNodeInfo()
```
Comment (llali) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/ObjectExplorer/ObjectExplorerServiceTests.cs]: please also add test for when the database is specified or empty

**Comment by llali** [comment]
>                 Server server = ServerCreator.Create(connection);

I'm pretty sure this also has to change for database instance. the query context for database and setting the parent

---
Refers to: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/ServerNode.cs:168 in a9acbee. [](commit_id = a9acbee361dfb09a1856b01e825e75201fb2ffaa, deletion_comment = False)

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/ServerNode.cs [hunk 3]
```diff
@@ -47,10 +47,19 @@ public ServerNode(ConnectionCompleteParams connInfo, IMultiServiceProvider servi
 
             this.context = new Lazy<SmoQueryContext>(() => CreateContext(serviceProvider));
 
-            NodeValue = connectionSummary.ServerName;
+            if (IsMasterConnection())
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/ServerNode.cs]: I'm pretty sure there's something wrong with this approach. this root code is not really database instance node and there are some logic in the database node needed for expanding and scripting. The context for server node and database node are different and database has a smo object because it is scriptable. the tests should verify that

**Comment by llali** [comment]
:clock1:

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/ServerNode.cs [hunk 4]
```diff
@@ -70,6 +79,12 @@ internal SmoServerCreator ServerCreator
             }
         }
 
+        internal bool IsMasterConnection()
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/SmoModel/ServerNode.cs]: we probably should check if the use is connecting to any of the system tables. e.g., master, tempdb, msdb or model.

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/ObjectExplorerUtils.cs [hunk 5]
```diff
@@ -71,6 +72,15 @@ public static TreeNode FindNode(TreeNode node, Predicate<TreeNode> condition, Pr
                 }
             }
             return null;
+        }
+
+        public static bool IsSystemDatabaseConnection(string databaseName)
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/ObjectExplorerUtils.cs]: please add comment for public methods

File: src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/ObjectExplorerService.cs [hunk 6]
```diff
@@ -346,8 +346,19 @@ public ObjectExplorerSession(string uri, TreeNode root, IMultiServiceProvider se
...
             public static ObjectExplorerSession CreateSession(ConnectionCompleteParams response, IMultiServiceProvider serviceProvider)
             {
-                TreeNode serverNode = new ServerNode(response, serviceProvider);
-                return new ObjectExplorerSession(response.OwnerUri, serverNode, serviceProvider, serviceProvider.GetService<ConnectionService>());
+                TreeNode rootNode = new ServerNode(response, serviceProvider);
+                var session = new ObjectExplorerSession(response.OwnerUri, rootNode, serviceProvider, serviceProvider.GetService<ConnectionService>());
+                if (!ObjectExplorerUtils.IsSystemDatabaseConnection(response.ConnectionSummary.DatabaseName))
+                {
+                    var children = rootNode.Expand();
+                    var databasesRoot = children.FirstOrDefault(x => x.NodeTypeId == NodeTypes.Databases);
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/ObjectExplorerService.cs]: this code assuming the databases are in a folder under server. if the structure change the code will break. you might want to traverse the tree to find the database node

Comment (AbbiePetcht) [inline src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/ObjectExplorerService.cs]: Will write the comment on assumption. If we change the structure, it will break the OE tests

---
In reply to: [112340512](https://github.com/Microsoft/sqltoolsservice/pull/322#discussion_r112340512) [](ancestors = 112340512)

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/ObjectExplorer/ObjectExplorerServiceTests.cs (hunk 2 above)
Comment (AbbiePetcht) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/ObjectExplorer/ObjectExplorerServiceTests.cs]: Already added it in integration test

---
In reply to: [112104491](https://github.com/Microsoft/sqltoolsservice/pull/322#discussion_r112104491) [](ancestors = 112104491)

**Comment by Raymondd** [comment]
Appveyor says the build passed, but it looks like a test is still failing: https://ci.appveyor.com/project/kburtram/sqltoolsservice/build/1.0.911/tests
